### ISSUE TITLE
False negative in jjwt-none-alg

### ISSUE THREAD

**Issue by sarr424** [returntocorp/sgrep-rules#3759]
**Describe the bug**
A clear and concise description of what the bug is and for which rule. Is the rule producing a false positive? Is it incorrect or missing something?

This is for:
https://semgrep.dev/playground/r/java.jjwt.security.jwt-none-alg.jjwt-none-alg

There is a subtle problem in the whitelist pattern in the current implementation; actually, the rule cannot ensure `signWith()` is called after the `builder()`.
```
    patterns:
      - pattern: |
          io.jsonwebtoken.Jwts.builder();
      - pattern-not-inside: |-
          $RETURNTYPE $FUNC(...) {
            ...
            $JWTS.signWith(...);
            ...
          }
```

This approach only verifies that a `signWith()` call exists somewhere inside the same method. However, if the method contains conditional branches, a `builder()` call in a branch without `signWith()` may be incorrectly considered safe because a signWith() appears in another branch.

Consider the example below. In the if branch, signWith() is used, so that usage is safe. But the else branch also contains a `builder()` call without `signWith()`. Because the pattern-not-inside matches the presence of `signWith()` anywhere in the method, it filters out the dangerous `builder()` call as well, resulting in a false negative.
```
public class App
{

    private static void bad1() {
        String jws;
        if (shouldSign) {
            Key key = Keys.secretKeyFor(SignatureAlgorithm.HS256);
            jws = Jwts.builder()
                    .setSubject("Bob")
                    .signWith(key)           // ok for this branch
                    .compact();
        } else {
            // ruleid: jjwt-none-alg
            jws = Jwts.builder()           // we should alarm on this branch
                    .setSubject("Bob")
                    .compact();
        }
    }
}
```

**To Reproduce**
Sample code to reproduce this behavior.

https://semgrep.dev/playground/s/Rr4JO

**Expected behavior**

The following pattern works:
```
    patterns:
      - pattern: |
          io.jsonwebtoken.Jwts.builder();
      - pattern-not-inside: |
          io.jsonwebtoken.Jwts.builder().....signWith(...);
```

**Priority**
How important is this to you?
- [ ] P0: blocking me from making progress
- [x] P1: this will block me in the near future
- [ ] P2: annoying but not blocking me

**Additional Context**

**Comment by fr4nsyz** [comment]
Hey I'm not seeing activity in this repo since March, so if it's alright, I'd like to pick this up.

I'll just reiterate my understanding of the problem to make sure:

`
The rule's pattern-not-inside checks for signWith() at the method scope, so a builder() call in one branch gets incorrectly treated as safe if signWith() exists in another branch of the same method.
`

So my proposed fix is to:

`Replace the method-scoped pattern-not-inside with a chain-scoped deep ellipsis pattern (Jwts.builder().....signWith(...)) that only excludes the match when signWith() is chained on the same builder() call.`

I also added a regression test (bad2_conditional) covering the conditional branch case.

Do you think this up to snuff?

https://github.com/semgrep/semgrep-rules/pull/3854
